lib/new.js

```diff
@@ -0,0 +1,687 @@
+/**
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const path = require('path');
+const puppeteer = require('puppeteer-core');const findChrome = require('./find_chrome');
+const {rpc} = require('../rpc');
+const debugApp = require('debug')('carlo:app');
+const debugServer = require('debug')('carlo:server');
+const {Color} = require('./color');
+const {HttpRequest} = require('./http_request');
+
+const fs = require('fs');
+const util = require('util');
+const {URL} = require('url');
+const EventEmitter = require('events');
+const fsReadFile = util.promisify(fs.readFile);
+
+let testMode = false;
+
+class App extends EventEmitter {
+  /**
+   * @param {!Puppeteer.Browser} browser Puppeteer browser
+   * @param {!Object} options
+   */
+  constructor(browser, options) {
```

Comment:
format: style mismatch:
```javascript
39|   */
40|  constructor(browser, options) {
41|    super();
```
```suggestion
  constructor(browser, options) {super();this.browser_ = browser;this.options_ = options;    this.windows_ = new Map();    this.exposedFunctions_ = [];this.pendingWindows_ = new Map();this.windowSeq_ = 0;this.www_ = [];}
```

⏎␣⁺␣⁺ at column 34 should be removed.
Triggered rule # 284
```
  +1.roles not in {ARGUMENT, CALLEE, THIS}
	∧ +1.reserved not in {=, ), {}
	∧ +2.length ≤ 2
	∧ +2.reserved not in {(}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.roles not in {IDENTIFIER}
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ -4.diff_col ≥ 6
	∧ •••start_col ≥ 18
	⇒ y = ⏎␣⁻␣⁻
	Confidence: 0.567. Support: 15.
```
⏎ at column 13 should be removed.
Triggered rule # 274
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.roles not in {RIGHT}
	∧ +2.reserved = (
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	⇒ y = ⏎
	Confidence: 0.837. Support: 46.
```
⏎ at column 29 should be removed.
Triggered rule # 290
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles not in {BLOCK, IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.internal_type = ClassDeclaration
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ -5.diff_offset ≤ 37
	⇒ y = ⏎␣⁻␣⁻
	Confidence: 0.700. Support: 25.
```
⏎ at column 91 should be removed.
Triggered rule # 111
```
  +1.roles in {EXPRESSION} and not in {LITERAL}
	∧ +2.roles not in {RIGHT}
	∧ +3.roles in {EXPRESSION} and not in {CALLEE}
	∧ +5.internal_type not in {Identifier}
	∧ ^1.roles not in {SCOPE, DECLARATION}
	∧ ^2.internal_type not in {File}
	∧ -1.length ≤ 2
	∧ -1.reserved not in {:, .}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -2.diff_line ≥ 1
	∧ -5.diff_line ≥ 2
	∧ •••start_line ≥ 34
	⇒ y = ∅
	Confidence: 0.962. Support: 13.
```
⏎ at column 38 should be removed.
Triggered rule # 288
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.length ≤ 2
	∧ +2.reserved not in {(}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ •••start_col ≤ 17
	⇒ y = ∅
	Confidence: 0.815. Support: 100.
```
⏎ at column 25 should be removed.
Triggered rule # 274
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.roles not in {RIGHT}
	∧ +2.reserved = (
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	⇒ y = ⏎
	Confidence: 0.837. Support: 46.
```
⏎␣⁻␣⁻ at column 20 should be removed.
Triggered rule # 122
```
  +1.roles in {ARGUMENT}
	∧ +2.roles not in {RIGHT}
	∧ +4.reserved not in {const}
	∧ ^1.roles not in {DECLARATION}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {:, .}
	∧ -2.diff_line = 0
	∧ -3.length ≤ 8
	∧ -3.diff_offset ≥ 4
	∧ -3.label in {<space>}
	∧ -4.reserved = =
	⇒ y = '
	Confidence: 0.500. Support: 19.
```